plugin/httptrace/clienttrace.go

```diff
@@ -45,112 +44,124 @@ type clientTracer struct {
-	_, sp := ct.tr.Start(ct.Context, name, trace.WithAttributes(attrs...), trace.WithSpanKind(trace.SpanKindClient))
+func (ct *clientTracer) start(hook, span string, attrs ...core.KeyValue) {
+	_, sp := ct.tr.Start(ct.Context, span, trace.WithAttributes(attrs...), trace.WithSpanKind(trace.SpanKindClient))
+	// TODO(paivagustavo): remove this for loop when `trace.WithAttributes(attrs...)` works.
+	for _, attr := range attrs {
+		sp.SetAttribute(attr)
+	}
 	ct.mtx.Lock()
 	defer ct.mtx.Unlock()
 	if ct.root == nil {
 		ct.root = sp
 	}
-	ct.levels[name] = sp
+	if _, ok := ct.activeHooks[hook]; ok {
+		// end was called before start is handled.
+		sp.End()
+		delete(ct.activeHooks, hook)
+	} else {
+		ct.activeHooks[hook] = sp
+	}
 }
 
-func (ct *clientTracer) close(name string) {
+func (ct *clientTracer) end(hook string) {
 	ct.mtx.Lock()
 	defer ct.mtx.Unlock()
-	if s, ok := ct.levels[name]; ok {
+	if s, ok := ct.activeHooks[hook]; ok {
 		s.End()
-		delete(ct.levels, name)
+		delete(ct.activeHooks, hook)
 	} else {
-		panic(fmt.Sprintf("failed to find span %s in levels.", name))
+		// start is not finished before end is called.
+		ct.activeHooks[hook] = trace.NoopSpan{}
 	}
 }
 
-func (ct *clientTracer) span(name string) trace.Span {
+func (ct *clientTracer) span(hook string) trace.Span {
 	ct.mtx.Lock()
 	defer ct.mtx.Unlock()
-	return ct.levels[name]
+	return ct.activeHooks[hook]
 }
 
 func (ct *clientTracer) getConn(host string) {
-	ct.open("http.getconn", HostKey.String(host))
+	ct.start("http.getconn", "http.getconn", HostKey.String(host))
 }
 
 func (ct *clientTracer) gotConn(info httptrace.GotConnInfo) {
-	ct.span("http.getconn").SetAttribute(HTTPRemoteAddr.String(info.Conn.RemoteAddr().String()))
-	ct.span("http.getconn").SetAttribute(HTTPLocalAddr.String(info.Conn.LocalAddr().String()))
-
-	ct.close("http.getconn")
+	ct.span("http.getconn").SetAttributes(
+		HTTPRemoteAddr.String(info.Conn.RemoteAddr().String()),
+		HTTPLocalAddr.String(info.Conn.LocalAddr().String()),
+	)
+	ct.end("http.getconn")
 }
 
 func (ct *clientTracer) putIdleConn(err error) {
 	if err != nil {
 		ct.span("http.receive").SetAttribute(MessageKey.String(err.Error()))
 		ct.span("http.receive").SetStatus(codes.Unknown)
 	}
-	ct.close("http.receive")
+	ct.end("http.receive")
 }
 
 func (ct *clientTracer) gotFirstResponseByte() {
-	ct.open("http.receive")
+	ct.start("http.receive", "http.receive")
 }
 
 func (ct *clientTracer) dnsStart(httptrace.DNSStartInfo) {
-	ct.open("http.dns")
+	ct.start("http.dns", "http.dns")
 }
 
 func (ct *clientTracer) dnsDone(httptrace.DNSDoneInfo) {
-	ct.close("http.dns")
+	ct.end("http.dns")
 }
 
 func (ct *clientTracer) connectStart(network, addr string) {
-	ct.open("http.connect")
+	ct.start("http.connect."+addr, "http.connect", HTTPRemoteAddr.String(addr))
 }
 
 func (ct *clientTracer) connectDone(network, addr string, err error) {
-	ct.close("http.connect")
+	ct.end("http.connect." + addr)
```

Comment:
It has always felt strange to me that we don't have the option to pass the `err` as an attribute to end to set it on that span as we close it...

but I guess we currently deal with this by calling `ct.span("http.getconn").SetAttributes(...)` before calling `ct.end()` (which also has its own way of blowing up on the outside case that start/end timing is reversed. shouldn't happen elsewhere in the `http` code as far as I know, buuut...)

Reply:
I've added an error and a list of attributes to be added when `end` is called and now only `start` and `end` modifies children spans, this should be enough to handle any race condition on `start`/`end` methods.